### PR TITLE
Only call +before/afterEach on CDRHooks conformers

### PR DESCRIPTION
**Pull request by idoru**
Rebased from #365, some small refactoring and addition of tests for `+afterEach` hook.

Tests for `+afterEach` are performed via asserts that are set up using `atexit(3)` since Cedar specs don't have visibility outside of their own context. Would like to know if anyone has better ideas for how to test this.

@tjarratt and I debated the inclusion of the test classes in `CDRHooksSpec.mm` but we ultimately felt it made the distinguishing characteristics clear. Curious to know how others feel about it.

This change also stops special casing for `UIAccessibilitySafeCategory__NSObject` and `SCRCException` since we shouldn't expect them to conform to `CDRHooks`.

### CODE REVIEW

File: Spec/CDRHooksSpec.mm [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+describe(@"global beforeEach", ^{
+    it(@"should run the +beforeEach before only for CDRHooks conformant classes", ^{
+        expect(conformantClassBeforeEachWasTriggered__).to(be_truthy);
+        expect(nonConformantClassBeforeEachWasTriggered__).to(be_falsy);
+    });
+});
+
+describe(@"global afterEach", ^{
+    it(@"should run after all specs", ^{
+        atexit(verifyAfterEachCalledAsExpected);
```
Comment (tjarratt) [inline Spec/CDRHooksSpec.mm]: This is awesome! I'll have to remember this trick.

**Comment by tjarratt** [comment]
Interesting point about not having to special case those classes now -- because that if check should bail out early when those classes don't conform to our protocol, it should be safe to call that function with any `Class`. That's a great improvement to this bit of code.

File: Spec/CDRHooksSpec.mm (hunk 1 above)
Comment (idoru) [inline Spec/CDRHooksSpec.mm]: This probably further warrants that the two class examples are kept here and only have effect on  static variables local to the test implementation - that way no other machinery (including other tests) can unintentionally mess with this state between the time these tests run and the time the process exits and triggers the `NSCAssert` validations.

**Comment by idoru** [comment]
Having slept on this, I'm not happy that these tests only really can see if the methods were called, but not when.

In fact, as soon as a single example runs (quite likely before this test), then `+beforeEach` and `+afterEach` on the conformant class will already have been called.

**Comment by idoru** [comment]
2db594b is closer to what I'm thinking but I only realized when I finally ran rake that this doesn't work on suites because the TestObservationHelper relies on XCTest.

File: Spec/SpecSpec.mm [hunk 2]
```diff
@@ -205,7 +205,6 @@ void expectFailure(CDRSpecBlock block) {
             globalValue__ = @"something";
         });
 
-        expect(globalValue__).to(be_nil);
```
Comment (idoru) [inline Spec/SpecSpec.mm]: @briancroom I think this was added in error 2 years ago... What's odd is that the example further up on line 174 seems to imply this should have raised an exception in the main spec run, but it only occurred when the spec gets redefined via construction of a `CDRSpecRun` in `CDRHooksSpec`.

**Pull request merged**